internal/controllers/topology/cluster/reconcile_state.go

```diff
@@ -721,14 +731,218 @@ func (r *Reconciler) deleteMachineDeployment(ctx context.Context, cluster *clust
+	// Do not create the MachinePool if it is marked as pending create.
+	// This will also block MHC creation because creating the MHC without the corresponding
+	// MachinePool is unnecessary.
+	mpTopologyName, ok := mp.Object.Labels[clusterv1.ClusterTopologyMachinePoolNameLabel]
+	if !ok || mpTopologyName == "" {
+		// Note: This is only an additional safety check and should not happen. The label will always be added when computing
+		// the desired MachinePool.
+		return errors.Errorf("new MachinePool is missing the %q label", clusterv1.ClusterTopologyMachinePoolNameLabel)
+	}
+	// Return early if the MachinePool is pending create.
+	if s.UpgradeTracker.MachinePools.IsPendingCreate(mpTopologyName) {
+		return nil
+	}
+
+	log := tlog.LoggerFrom(ctx).WithMachinePool(mp.Object)
+	cluster := s.Current.Cluster
+	infraCtx, _ := log.WithObject(mp.InfrastructureMachinePoolObject).Into(ctx)
+	if err := r.reconcileReferencedObject(infraCtx, reconcileReferencedObjectInput{
+		cluster: cluster,
+		desired: mp.InfrastructureMachinePoolObject,
+	}); err != nil {
+		return errors.Wrapf(err, "failed to create %s", mp.Object.Kind)
+	}
+
+	bootstrapCtx, _ := log.WithObject(mp.BootstrapObject).Into(ctx)
+	if err := r.reconcileReferencedObject(bootstrapCtx, reconcileReferencedObjectInput{
+		cluster: cluster,
+		desired: mp.BootstrapObject,
+	}); err != nil {
+		return errors.Wrapf(err, "failed to create %s", mp.Object.Kind)
+	}
+
+	log = log.WithObject(mp.Object)
+	log.Infof(fmt.Sprintf("Creating %s", tlog.KObj{Obj: mp.Object}))
+	helper, err := r.patchHelperFactory(ctx, nil, mp.Object)
+	if err != nil {
+		return createErrorWithoutObjectName(ctx, err, mp.Object)
+	}
+	if err := helper.Patch(ctx); err != nil {
+		return createErrorWithoutObjectName(ctx, err, mp.Object)
+	}
+	r.recorder.Eventf(cluster, corev1.EventTypeNormal, createEventReason, "Created %q", tlog.KObj{Obj: mp.Object})
+
+	// Wait until MachinePool is visible in the cache.
+	// Note: We have to do this because otherwise using a cached client in current state could
+	// miss a newly created MachinePool (because the cache might be stale).
+	err = wait.PollUntilContextTimeout(ctx, 5*time.Millisecond, 5*time.Second, true, func(ctx context.Context) (bool, error) {
+		key := client.ObjectKey{Namespace: mp.Object.Namespace, Name: mp.Object.Name}
+		if err := r.Client.Get(ctx, key, &expv1.MachinePool{}); err != nil {
+			if apierrors.IsNotFound(err) {
+				return false, nil
+			}
+			return false, err
+		}
+		return true, nil
+	})
+	if err != nil {
+		return errors.Wrapf(err, "failed waiting for MachinePool %s to be visible in the cache after create", mp.Object.Kind)
+	}
+
+	return nil
+}
+
+// updateMachinePool updates a MachinePool. Also rotates the corresponding Templates if necessary.
+func (r *Reconciler) updateMachinePool(ctx context.Context, cluster *clusterv1.Cluster, currentMP, desiredMP *scope.MachinePoolState) error {
+	log := tlog.LoggerFrom(ctx).WithMachinePool(desiredMP.Object)
+
```

Comment:
The return early for `IsPendingUpgrade` is missing here (see updateMachinePool)

```go
	// Return early if the MachinePool is pending an upgrade.
	// Do not reconcile the MachinePool yet to avoid updating the MachinePool while it is still pending a
	// version upgrade. This will prevent the MachinePool from performing a double rollout.
	if s.UpgradeTracker.MachinePools.IsPendingUpgrade(currentMP.Object.Name) {
		return nil
	}
```